keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md

```diff
@@ -0,0 +1,163 @@
+---
+kep-number: 0
+title: Bounding Self-Labeling Kubelets
+authors:
+  - "@mikedanese"
+  - "@liggitt"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-node
+  - sig-storage
+reviewers:
+  - "@saad-ali"
+  - "@tallclair"
+  - "@deads2k"
+approvers:
+  - "@tallclair"
+  - "@smarterclayton"
+creation-date: 2017-08-14
+last-updated: 2018-10-25
+status: provisional
+---
+
+# Bounding Self-Labeling Kubelets
+
+## Motivation
+
+Today the node client has total authority over its own Node labels.
+This ability is incredibly useful for the node auto-registration flow.
+The kubelet reports a set of well-known labels, as well as additional
+labels specified on the command line with `--node-labels`.
+
+While this distributed method of registration is convenient and expedient, it
+has two problems that a centralized approach would not have. Minorly, it makes
+management difficult. Instead of configuring labels in a centralized
+place, we must configure `N` kubelet command lines. More significantly, the
+approach greatly compromises security. Below are two straightforward escalations
+on an initially compromised node that exhibit the attack vector.
+
+### Capturing Dedicated Workloads
+
+Suppose company `foo` needs to run an application that deals with PII on
+dedicated nodes to comply with government regulation. A common mechanism for
+implementing dedicated nodes in Kubernetes today is to set a label or taint
+(e.g. `foo/dedicated=customer-info-app`) on the node and to select these
+dedicated nodes in the workload controller running `customer-info-app`.
+
+Since the nodes self reports labels upon registration, an intruder can easily
+register a compromised node with label `foo/dedicated=customer-info-app`. The
+scheduler will then bind `customer-info-app` to the compromised node potentially
+giving the intruder easy access to the PII.
+
+This attack also extends to secrets. Suppose company `foo` runs their outward
+facing nginx on dedicated nodes to reduce exposure to the company's publicly
+trusted server certificates. They use the secret mechanism to distribute the
+serving certificate key. An intruder captures the dedicated nginx workload in
+the same way and can now use the node certificate to read the company's serving
+certificate key.
+
+## Proposal
+
+The NodeRestriction admission plugin will be updated to limit the labels a kubelet
+can set on its Node object. Kubelets will only be allowed to set the following labels.
```

Comment:
I think we use the node taint to make sure application pods can not run before all base node containers are running to fix for example kube2iam not running before an application that needs aws api access granted by the role fetched by kube2iam.
I think somehow we need a way how to mark a node to be not ready for application containers to run to help our teams to have no weird problems occurring.

@mikkeloscar can you check if our use case of bootstrapping nodes are violated by this change?

Reply:
We don't currently have this in place as it didn't work with the autoscaler.

However we use the label feature a lot to identify different node pools in our clusters and this we do by passing the `--node-labels` flag. While I can understand the security concern in certain environments I don't see the reason for removing a useful feature rather than just make it possible to restrict it if needed. Simply make it configurable on the apiserver what labels are allowed and let the cluster admins decide whether they want this protection or not. I'm fine with making it a default as long as we can disable it.

Assuming this would be in place and `--node-labels` would be dropped, how is it intended that we label nodes in an easy way?

Also the argument about making management difficult is a bit far fetched IMO. You don't configure `N` kubelet command lines, you have the configuration somewhere and reuse it for all nodes that are supposed to have the same configuration. [This is how we do it](https://github.com/zalando-incubator/kubernetes-on-aws/blob/bc59ef967e9d78358ac6bfb80c75525cf51b64a3/cluster/node-pools/worker-default/userdata.clc.yaml#L162). We manage 100's of nodes like this, it's not an issue :)

Reply:
> @szuecs:
> I think we use the node taint to make sure application pods can not run before all base node containers are running to fix for example kube2iam not running before an application that needs aws api access granted by the role fetched by kube2iam. I think somehow we need a way how to mark a node to be not ready for application containers to run to help our teams to have no weird problems occurring.

This is not removing the ability of nodes to add taints on creation.

> @mikkeloscar:
> While I can understand the security concern in certain environments I don't see the reason for removing a useful feature rather than just make it possible to restrict it if needed. Simply make it configurable on the apiserver what labels are allowed and let the cluster admins decide whether they want this protection or not. I'm fine with making it a default as long as we can disable it.

On the apiserver side, all of the proposed restrictions are within the [NodeRestriction](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) admission plugin. The job of this plugin is to restrict nodes to the workloads that are scheduled to run on them. `kube-apiserver` does *not enable it by default*. If you chose to run without that admission plugin enabled, your nodes could continue to label themselves arbitrarily.

That said, it is important for the kubelet to work well with recommended apiserver configurations, which is why this is proposing limiting the labels `--node-labels` accepts to ones the NodeRestriction plugin would allow.

> Assuming this would be in place and --node-labels would be dropped, how is it intended that we label nodes in an easy way?

This is not dropping `--node-labels`, it is limiting the set of labels to ones that can be distinguished as kubelet-set labels, like `kubelet.kubernetes.io/foo`. Over the next two releases, kubelet invocations would need to move to labels with those prefixes.

Reply:
Thanks for your clarification

Reply:
@liggitt
> Over the next two releases, kubelet invocations using --node-labels would need to move to labels with those prefixes.

Would this still only be the case if `NodeRestriction` plugin is enabled, or would it be enforced on the `kubelet` side? If it's enforced by the kubelet it will be a bit annoying.